### DISCUSSION TITLE
Save Ram Conversion Utility

### DISCUSSION THREAD

**Discussion by Cephiros** [sanni/cartreader#593] [Look what I made]
Here is a quick program  i wrote to convert saves between the following types, specifically.
MD/Genesis: OSCR sram dump to kega fusion and back. 
N64: OSCR sram, and flash dump to Project64 and back. 
Gameboy/Color: OSCR MBC2 sram dump to mGBA and back. 

Will likely work with other emulators, But I've only tested with what I use to test dumps.

![screenshot](https://user-images.githubusercontent.com/9114282/198495226-bd3c6898-56b5-4182-b413-e468892e1fcf.png)
[Open Source Cartridge Reader Conversions.zip](https://github.com/sanni/cartreader/files/10670369/Open.Source.Cartridge.Reader.Conversions.zip)

Updated: Fixed bug preventing correct project64 conversion.

**Comment by sebala82** [comment]
Hey @Cephiros!
First of all thanks for sharing this awesome tool (although I'm having trouble getting it to work, as I'm about to explain now)...
I dumped the roms and the saves of both my Ocarina of Time PAL(v1.0) and NTSC-J(v1.0) copies using sanni's OSCR.
Trying to check, if I could use the saves in an emulator, I was informed, that the file formats aren't compatible as is and that you wrote an awesome tool to convert between the two. Unfortunately both converted savefiles aren't accepted by project64 (current version).
Maybe you (or someone else with experience with sra-files) could take a look at the files? It would be greatly appreciated! (I obviously cannot upload my rom dumps here, if anyone needs them for testing, please let me know!)
Thanks alot in advance!
Sebastian
[OOT_Savefiles.zip](https://github.com/sanni/cartreader/files/10668691/OOT_Savefiles.zip)

**Comment by Cephiros** [comment]
You are correct something strange is going on, When I get home I will have to look into the code and see if there is a typo or change in the final compile that wasnt there in testing, if you look at the headers in the original its` �˜.!ZELDA`, its supposed to convert to `˜�Z!.ADLE` However for some reason the output is `˜˜Z!Z.ADAE`, its repeating the first byte in the third position in each 4 byte block, instead of the expected 3rd byte.

**Reply by sebala82** [reply]
I noticed this aswell but I have no clue whatsoever of how this file format works ;-)
Thanks a ton for your help!

**Reply by Cephiros** [reply]
Pretty simple really, lest say the save dumped like "1234123412341234", for unknown reasons project64, and possibly other emulators expect the data in the format of "4321432143214321". Stranger yet this doesnt effect all save types, Specifically EEPROM saves read normally from what I have tested..

**Reply by Cephiros** [reply]
Ive found, and fixed the problem. Go ahead and try the new version in the original post.

Sorry about that.

**Reply by sebala82** [reply]
It works now!
Thank you so much!

**Comment by asveikau** [comment]
Hi @Cephiros , I'm just a random onlooker, but given that this is github, have you considered creating a github repo for this tool and adding C# source?  Not meaning to tell you what to do but I suspect it may be of more use to the community that way, vs. distributing it as a compiled binary on a discussion thread.  Looks like a cool tool.  Cheers.

**Reply by Cephiros** [reply]
I understand this, but truth be told, the code is quickly thrown together(messy) with no notations, its not anything i would feel I want to expose to the world. I made it for me, but posted it in the off chance others can use it. But in any case, I believe everything in there accept the gameboy fix is already elsewhere in code. I will say, if you are really curious, It is not obfuscated in any way. I dont feel a need to protect my code, I just dont think its personally presentable.

**Comment by b3nner** [comment]
Can this convert snes srm files from retroarch back onto original carts via the oscr? I am having issues with exporting srm files made on retroarch back onto snes carts, i get verification errors.
